firewood/src/storage/linear/mod.rs

```diff
@@ -0,0 +1,98 @@
+//! stateDiagram-v2
+//!     R1(Committed) --> R2(Committed)
+//!     R2(Committed) --> R3(FileBacked)
+//!     P1(Proposed) --> R3(FileBacked)
+//!     P2(Proposed) --> P1(Proposed)
+//! ```
+//!
+//! Each type is described in more detail below.
+
+use std::fmt::Debug;
+use std::io::{Error, Read};
+
+mod committed;
+/// A linear store used for proposals
+///
+/// A Proposed LinearStore supports read operations which look for the
+/// changed bytes in the `new` member, and if not present, delegate to
+/// their parent.
+///
+/// The old map is maintained because it is needed if this proposal commits.
+/// The old and new maps contain the same entries unless the address extends
+/// beyond the end of the base linear store, in which case only new bytes
+/// exist
+///
+/// The parent can either be another `Proposed` store or a `FileBacked` store.
+///
+/// The second generic parameter specifies whether this proposal is mutable or
+/// not. Mutable proposals implement the `ReadWriteLinearStore` trait
+///
+/// The possible combinations are:
+///  - `Proposed<Proposed, ReadWrite>` (in-progress nested proposal)
+///  - `Proposed<Proposed, ReadOnly>` (completed nested proposal)
+///  - `Proposed<FileBacked, ReadWrite>` (first proposal on base revision)
+///  - `Proposed<FileBacked, ReadOnly>` (completed first proposal on base)
+///
+/// Transitioning from ReadWrite to ReadOnly just prevents future mutations to
+/// the proposal maps. ReadWrite proposals only exist during the application of
+/// a Batch to the proposal, and are subsequently changed to ReadOnly
+///
+/// # How a commit works
+///
+/// Lets assume we have the following:
+///  - bytes "on disk":   (0, 1, 2) `LinearStore<FileBacked>`
+///  - bytes in proposal: (   3   ) `LinearStore<Proposed<FileBacked, ReadOnly>>`
+/// that is, we're changing the second byte (1) to (3)
+///
+/// To commit:
+///  - Convert the `LinearStore<FileBacked>` to `LinearStore<Committed>` taking the
+/// old pages from the `LinearStore<Proposed<FileBacked, Readonly>>`
+///  - Change any direct child proposals from `LinearStore<Proposed<Proposed, Readonly>>`
+/// into `LinearStore<FileBacked>`
+///  - Invalidate any other `LinearStore` that is a child of `LinearStore<FileBacked>`
+///  - Flush all the `Proposed<FileBacked, ReadOnly>::new` bytes to disk
+///  - Convert the `LinearStore<Proposed<FileBacked, Readonly>>` to `LinearStore<FileBacked>`
+mod filebacked;
+mod proposed;
+
+#[derive(Debug)]
+pub(super) struct LinearStore<S: ReadLinearStore> {
+    state: S,
+}
+
+/// All linearstores support reads
+pub(super) trait ReadLinearStore: Debug {
+    fn stream_from(&self, addr: u64) -> Result<impl Read, Error>;
+    fn size(&self) -> Result<u64, Error>;
+}
+
+/// Some linear stores support updates
+pub(super) trait WriteLinearStore: Debug {
+    fn write(&mut self, offset: u64, object: &[u8]) -> Result<usize, Error>;
+}
+
+impl<ReadWrite: ReadLinearStore + Debug> WriteLinearStore for LinearStore<ReadWrite> {
```

Comment:
Should this be `ReadWrite: ReadLinearStore + WriteLinearStore + Debug`?
```suggestion
impl<ReadWrite: ReadLinearStore + Debug> WriteLinearStore for LinearStore<ReadWrite> {
```

Reply:
> Should this be `ReadWrite: ReadLinearStore + WriteLinearStore + Debug`?
Yes. Fixed when these methods were implemented in an upcoming diff.

Reply:
Does it ever make sense to have a linear store that's `Read` but not `Write`? Can simplify your trait bounds by making `trait WriteLinearStore: ReadLinearStore { ... }`

Reply:
Fixed in newest code